main/broadcasters.c

```diff
@@ -389,12 +388,78 @@ static broadcaster_ops_t mijia_temp_hum_ops = {
     .is_broadcaster = mijia_temp_hum_is_broadcaster,
     .metadata_get = mijia_temp_hum_metadata_get,
 };
+/* Beewi Smart Door */
+#define BEEWI_SMART_DOOR_COMPANY_ID     0x000D
+#define BEEWI_SMART_DOOR_SERVICE_ID     0x08 
+#define BEEWI_SMART_DOOR_DATA_TYPE_STAT 0x0C
+#define BEEWI_SMART_DOOR_DATA_TYPE_BATT 0x06
+
+
+typedef struct {
+    uint16_t company_id;
+    uint8_t  service_id;
+	uint8_t  type_stat;
+    uint8_t  status;
+	uint8_t  type_batt;
+	uint8_t  battery;
+} __attribute__((packed)) beewi_smart_door_t;
```

Comment:
Are you sure this is the exact format of the payload?
Could it be that the status and battery fields may either be included or not and according to the type field you should check?
Do you have any documentation you can share regarding the advertisement payload?

Reply:
No doc at all they refused to communicate specs. This is just reverse enginering.
Status and Batt are always present in this frame.

Here is a complete packet : 
02 01 06 02 0A 00 08 FF 0D 00 08 0C **01** 06 **64** 11 09 42 65 57 69 20 53 6D 61 72 74 20 44 6F 6F 72 => Open
02 01 06 02 0A 00 08 FF 0D 00 08 0C **00** 06 **64** 11 09 42 65 57 69 20 53 6D 61 72 74 20 44 6F 6F 72 => Closed

As there's always 0C before Status and 06 before Battery I've defined them as identifier.
0D 00 part before is the Company ID ( not from the device but from the bt chipset i to the standard, Texas Instrument here ( 00 0D ) .
Here is complete wireshark analysis of the packet : 

`Frame 435: 44 bytes on wire (352 bits), 44 bytes captured (352 bits)
    Encapsulation type: Bluetooth H4 with linux header (99)
    Arrival Time: Jan 22, 2019 11:33:07.735169000 Paris, Madrid
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1548153187.735169000 seconds
    [Time delta from previous captured frame: 0.259482000 seconds]
    [Time delta from previous displayed frame: 0.259482000 seconds]
    [Time since reference or first frame: 41.804692000 seconds]
    Frame Number: 435
    Frame Length: 44 bytes (352 bits)
    Capture Length: 44 bytes (352 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    Point-to-Point Direction: Received (1)
    [Protocols in frame: bluetooth:hci_h4:bthci_evt:btcommon]
Bluetooth
    [Source: controller]
    [Destination: host]
Bluetooth HCI H4
    [Direction: Rcvd (0x01)]
    HCI Packet Type: HCI Event (0x04)
Bluetooth HCI Event - LE Meta
    Event Code: LE Meta (0x3e)
    Parameter Total Length: 41
    Sub Event: LE Extended Advertising Report (0x0d)
    Num Reports: 1
    Event Type: 0x0013, Connectable, Scannable, Legacy, Data Status: Complete
    Peer Address Type: Public Device Address (0x00)
    BD_ADDR: TexasIns_e3:a4:e6 (10:ce:a9:e3:a4:e6)
    Primary PHY: LE 1M (0x01)
    Secondary PHY: No packets on the secondary advertising channel (0x00)
    Advertising SID: 0xff (not available)
    TX Power: 127dBm (not available)
    RSSI: -65dBm
    Periodic Advertising Interval: 0x0000 (no periodic advertising)
    Direct Address Type: Public Device Address (0x00)
    Direct BD_ADDR: 00:00:00_00:00:00 (00:00:00:00:00:00)
    Data Length: 15
    Advertising Data
        Flags
            Length: 2
            Type: Flags (0x01)
            000. .... = Reserved: 0x0
            ...0 .... = Simultaneous LE and BR/EDR to Same Device Capable (Host): false (0x0)
            .... 0... = Simultaneous LE and BR/EDR to Same Device Capable (Controller): false (0x0)
            .... .1.. = BR/EDR Not Supported: true (0x1)
            .... ..1. = LE General Discoverable Mode: true (0x1)
            .... ...0 = LE Limited Discoverable Mode: false (0x0)
        Tx Power Level
            Length: 2
            Type: Tx Power Level (0x0a)
            Power Level (dBm): 0
        Manufacturer Specific
            Length: 8
            Type: Manufacturer Specific (0xff)
            Company ID: Texas Instruments Inc. (0x000d)
            Data: 080c010664
                [Expert Info (Note/Undecoded): Undecoded]
                    [Undecoded]
                    [Severity level: Note]
                    [Group: Undecoded]`

Reply:
OK, in that case, using the company ID to distinguish these devices isn't good enough.
I suggest you try to read the `ESP_BLE_AD_TYPE_NAME_CMPL` part of the advertisement data and compare it with `BeWi Smart Door`.
That should be specific enough. I would also suggest marking the `type_stat` and `type_batt` fields as `tbd1` and `tbd2` as we don't really know yet what they are.
Since we know for sure this is the smart door sensor (as we read its name), we can forgo checking these stat fields and just assume the values are where we expect them to be.

Reply:
I'll compare with advertised device name. That's what I'm already doing in another project.
But we also nee to check the type_stat as device broadcast another frame starting also with 00 D0 08 when in pairing mode. Couldn't figure what data there's in. Probably only an identifier string to say it's in pairing state.

Reply:
I'm trying this but it doesn't seems to work : 

`snprintf(dev_name, strlen(dev_name), "%s", esp_ble_resolve_adv_data(adv_data, ESP_BLE_AD_TYPE_NAME_CMPL, &dev_len ));
	
    if (strncmp(dev_name, "Beewi Smart Door", dev_len) == 0)
		return 1;
    else return 0;	`

Reply:
That's probably because the device sends `BeWi Smart Door` and not `BeeWi Smart Door`.
Also, `esp_ble_resolve_adv_data()` might return NULL so watch out for that. Since you're using `strncmp()` anyways, there really is no need to put it in a buffer first.